components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1467,6 +1470,36 @@ where
         }
     }
 
+    // If lease expired, we will send a noop read index to renew lease.
+    fn try_renew_leader_lease(&mut self) {
+        if !self.fsm.peer.is_leader() {
+            return;
+        }
+        if let Err(e) = self.fsm.peer.pre_read_index() {
+            debug!(
+                "prevent unsafe read index to renew leader lease";
+                "region_id" => self.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+                "err" => ?e,
+            );
+            self.ctx.raft_metrics.propose.unsafe_read_index += 1;
+            return;
+        }
+
+        let current_time = *self.ctx.current_time.get_or_insert_with(monotonic_raw_now);
+        let renew_bound = current_time + self.ctx.cfg.check_leader_lease_interval();
+        if self.fsm.peer.need_renew_lease_at(self.ctx, renew_bound) {
+            let (id, dropped) = self.fsm.peer.propose_read_index(None, None);
+            if dropped {
+                self.ctx.raft_metrics.propose.dropped_read_index += 1;
+                return;
+            }
+            self.ctx.raft_metrics.propose.read_index += 1;
+            let read_proposal = ReadIndexRequest::noop(id, current_time);
+            self.fsm.peer.pending_reads.push_back(read_proposal, true);
```

Comment:
Can these operations also be moved to `Peer` so that `pending_reads` doesn't have to be public?

Reply:
Any updates?